// ql-stats-model/lib/types/ModType.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isEnvironment = exports.ModType = void 0;
var ModType;
(function (ModType) {
    ModType["BFG"] = "BFG shot";
    ModType["BFG_SPLASH"] = "BFG splash";
    ModType["CHAINGUN"] = "Chain Gun";
    ModType["CRUSH"] = "Crush";
    /**
     * When the player was falling to death.
     */
    ModType["FALLING"] = "Falling";
    ModType["GAUNTLET"] = "Gauntlet slice";
    ModType["GRENADE"] = "Grenade";
    ModType["GRENADE_SPLASH"] = "Grenade splash";
    ModType["HMG"] = "Heavy Machine Gun bullet";
    /**
     * When the player falls out of the arena.
     */
    ModType["HURT"] = "Hurt";
    ModType["KAMIKAZE"] = "Kamikaze";
    ModType["LAVA"] = "Lava";
    ModType["LIGHTNING"] = "Lightning";
    ModType["MACHINEGUN"] = "Machine Gun bullet";
    ModType["NAILGUN"] = "Nail";
    ModType["PLASMA"] = "Plasma ball";
    ModType["PLASMA_SPLASH"] = "Plasma splash";
    ModType["PROXMINE"] = "Proximity Mine";
    ModType["PROXMINE_SPLASH"] = "Proximity Mine splash";
    ModType["RAILGUN"] = "Rail";
    ModType["ROCKET"] = "Rocket";
    ModType["ROCKET_SPLASH"] = "Rocket splash";
    ModType["SHOTGUN"] = "Shotgun bullet";
    ModType["SLIME"] = "Slime";
    ModType["SWITCHTEAM"] = "Switch team";
    ModType["TRIGGER_HURT"] = "Trigger hurt";
    ModType["WATER"] = "Water";
    /**
     * Unkwown world death
     */
    ModType["UNKNOWN"] = "Unknown";
})(ModType = exports.ModType || (exports.ModType = {}));
function isEnvironment(mod) {
    return mod == ModType.CRUSH ||
        mod == ModType.FALLING ||
        mod == ModType.HURT ||
        mod == ModType.LAVA ||
        mod == ModType.SLIME ||
        mod == ModType.TRIGGER_HURT ||
        mod == ModType.WATER ||
        mod == ModType.UNKNOWN;
}
exports.isEnvironment = isEnvironment;
